### DISCUSSION TITLE
tr/guides/object-counting/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8245] [Docs]
# tr/guides/object-counting/

Nesne Sayma Kullanımı Ultralytics YOLOv8

https://docs.ultralytics.com/tr/guides/object-counting/

**Comment by kkayaci** [comment]
Hello from Turkey,
There is more than one vehicle class (car, bus, truck, etc.) that we will count. If I want to track and count all of these classes separately, what kind of editing is needed for example 3 (specific classes) on this page? (Kenan Kayacı)

**Reply by pderrenger** [reply]
@kkayaci merhaba Kenan,

Türkiye'den selamlar! 🇹🇷 Araç sınıflarını (araba, otobüs, kamyon vb.) ayrı ayrı takip etmek ve saymak için, örnek 3'teki `classes_to_count` listesini düzenlemeniz gerekecek. İlgili sınıfların indekslerini bu listeye ekleyerek her birini ayrı ayrı sayabilirsiniz. İşte bir örnek:

```python
from ultralytics import YOLO
from ultralytics.solutions import object_counter
import cv2

# Load model
model = YOLO("yolov8n.pt")

# Video capture
cap = cv2.VideoCapture("path/to/video.mp4")
assert cap.isOpened(), "Error reading video file"

# Define classes to count (e.g., 2: 'car', 5: 'bus', 7: 'truck')
classes_to_count = [2, 5, 7]

# Initialize Object Counter
counter = object_counter.ObjectCounter()
counter.set_args(view_img=True, draw_tracks=True, classes_names=model.names)

# Process video
while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        break
    tracks = model.track(im0, persist=True, show=False, classes=classes_to_count)
    im0 = counter.start_counting(im0, tracks)
    # Display or save frame...

cap.release()
```

Bu kod, belirtilen sınıflar için nesne sayımını gerçekleştirecektir. Umarım bu yardımcı olur! Herhangi bir sorunuz olursa, lütfen sormaktan çekinmeyin. 😊

İyi çalışmalar!

**Reply by kkayaci** [reply]
Hello again Paula,
Thank you for your help. I did all the coding you gave. My goal is to separately count the value of cars, buses and trucks going left and right (or going up and down) passing through a given line or region. With this code, we can count the vehicles passing (entering and exiting) a certain line, regardless of the vehicle type. However, we cannot count vehicle types (car, bus and truck) separately. When we write classes_to_count=[2,5,7]; We track the vehicle types car, bus and truck, but it has not yet been possible to obtain the count of cars, buses and trucks going left and right (or up and down), what kind of coding do I need to do?

**Reply by pderrenger** [reply]
Merhaba Paula,

Yardımım işine yaradıysa ne mutlu bana! 🎉 Araç türlerini (araba, otobüs, kamyon) ayrı ayrı saymak için, her bir araç türünün geçiş yönünü takip eden ve sayan bir mantık eklemeniz gerekecek. `ObjectCounter` sınıfınızın her bir araç türü ve yön için ayrı sayımlar tutmasını sağlayacak şekilde genişletmeniz gerekebilir. İşte bir örnek kod parçası:

```python
from ultralytics import YOLO
from ultralytics.solutions import object_counter
import cv2

# Load model
model = YOLO("yolov8n.pt")

# Define classes to count
classes_to_count = [2, 5, 7]  # 2: 'car', 5: 'bus', 7: 'truck'

# Initialize Object Counter with separate counts for each direction
counter = object_counter.ObjectCounter()
counter.set_args(view_img=True, draw_tracks=True, classes_names=model.names)

# Define a function to determine the direction of travel
def determine_direction(track):
    # Implement logic to determine the direction based on track coordinates
    # Return 'left', 'right', 'up', or 'down'
    pass

# Process video
while True:
    # Read frame from video
    # ...

    # Run tracking
    tracks = model.track(frame, persist=True, show=False, classes=classes_to_count)

    # Count objects by type and direction
    for track in tracks:
        direction = determine_direction(track)
        obj_class = track.cls  # Object class
        counter.counts[obj_class][direction] += 1  # Increment count

    # Display or save frame with counts
    # ...
```

Bu kod, her bir araç türü için yön bazında sayım yapacak şekilde `ObjectCounter` sınıfını genişletir. `determine_direction` fonksiyonu, izlenen nesnenin yönünü belirlemek için koordinatları kullanır ve bu bilgiyi sayım için kullanır. Umarım bu yardımcı olur ve projenizde başarılar dilerim! 🚗🚌🚚

İyi kodlamalar!

**Reply by kkayaci** [reply]
Hello again,
After doing the necessary work, I will share the results. Thank you very much for your interest and help. Greetings and respect from Ankara.

**Reply by glenn-jocher** [reply]
@kkayaci merhaba,

Ankara'dan selamlar ve saygılar! 🌟 İlginiz ve desteğiniz için çok teşekkür ederim. Sonuçları paylaşmanızı sabırsızlıkla bekliyorum. Eğer yol boyunca herhangi bir yardıma ihtiyacınız olursa, lütfen çekinmeden bana ulaşın. 

Belirli sınıflar için nesne sayımı ve yön tespiti yapmak istediğinizde, her bir sınıf ve yön için ayrı sayımlar tutacak şekilde `ObjectCounter` sınıfını genişletmeniz gerekecek. İşte bir örnek:

```python
# ... (Önceki kodunuz)

# Araç sınıflarını ve yönlerini saymak için bir sözlük oluşturun
counts = {cls: {'left': 0, 'right': 0} for cls in classes_to_count}

# Her bir izi işleyin ve sayımları güncelleyin
for track in tracks:
    cls = track.cls  # Araç sınıfı
    direction = determine_direction(track)  # Araç yönünü belirleyin
    counts[cls][direction] += 1  # Sayımı güncelleyin

# Sayımları yazdırın veya kaydedin
for cls, directions in counts.items():
    print(f"Class {model.names[cls]} - Left: {directions['left']}, Right: {directions['right']}")
```

`determine_direction` fonksiyonu, izlenen nesnenin yönünü belirlemek için koordinatları kullanır ve bu bilgiyi sayım için kullanır. Bu fonksiyonun içeriğini, projenizin gereksinimlerine göre doldurmanız gerekecek.

Kolay gelsin ve başarılar dilerim!

**Reply by kkayaci** [reply]
Hello, I will implement the suggestions for my project and share them with you as soon as possible. Thank you for the support.

**Reply by glenn-jocher** [reply]
Merhaba,

Projende ilerleme kaydettiğine sevindim ve sonuçları bizimle paylaşacak olmandan dolayı heyecanlıyım! 🚀 Destek olabildiysem ne mutlu bana. Eğer yolculuğun sırasında herhangi bir sorunla karşılaşırsan, lütfen çekinmeden buradan sor. Ankara'dan selamlar ve saygılar! 

Mutlu kodlamalar! 😊👨‍💻

**Comment by kayahasan** [comment]
Greetings 
I try to count the pieces passing through the conveyor band with the weight I have prepared. However, when the products are flowing in one direction, 1 or 2 out of 15 products are strangely counted as out. How do I do one-way counting? 
Thank you

**Reply by pderrenger** [reply]
Greetings!

Thank you for reaching out. To achieve one-way counting on your conveyor belt, you can define a specific region or line for counting and ensure that the counting logic only considers objects moving in the desired direction. Here’s a concise example to help you set this up:

```python
import cv2
from ultralytics import YOLO, solutions

model = YOLO("yolov8n.pt")
cap = cv2.VideoCapture("path/to/video/file.mp4")
assert cap.isOpened(), "Error reading video file"
w, h, fps = (int(cap.get(x)) for x in (cv2.CAP_PROP_FRAME_WIDTH, cv2.CAP_PROP_FRAME_HEIGHT, cv2.CAP_PROP_FPS))

# Define line points for one-way counting
line_points = [(20, 400), (1080, 400)]

# Video writer
video_writer = cv2.VideoWriter("object_counting_output.avi", cv2.VideoWriter_fourcc(*"mp4v"), fps, (w, h))

# Init Object Counter
counter = solutions.ObjectCounter(
    view_img=True,
    reg_pts=line_points,
    classes_names=model.names,
    draw_tracks=True,
    line_thickness=2,
)

while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        print("Video frame is empty or video processing has been successfully completed.")
        break
    tracks = model.track(im0, persist=True, show=False)

    im0 = counter.start_counting(im0, tracks)
    video_writer.write(im0)

cap.release()
video_writer.release()
cv2.destroyAllWindows()
```

Ensure that the `line_points` are set correctly to define the counting line, and the `ObjectCounter` is configured to count objects crossing this line in the desired direction.

For more detailed guidance, you can refer to the [Object Counting Guide](https://docs.ultralytics.com/guides/object-counting/).

Best of luck with your project! If you have any further questions, feel free to ask. 😊

**Reply by kayahasan** [reply]
Hello,
Thank you for the help. I tried the code you gave, but my problem persists. I uploaded a video to YouTube for better understanding. While it is about to count the 13th product in the 15th second, it perceives the product going in the same direction as entering.

video: https://www.youtube.com/watch?v=05lcYl38P7c

**Reply by pderrenger** [reply]
@kayahasan hello,

Thank you for sharing the video; it helps clarify the issue. It seems like the counting anomaly might be due to the direction sensitivity of the counting logic or an overlap in the counting area where the product momentarily appears to move in an opposite direction due to perspective or object shape.

To ensure one-way counting and avoid miscounts, you can adjust the counting line or region to be more aligned with the flow direction. Here’s a suggestion:

1. **Adjust the Counting Line:** Ensure the counting line is perpendicular to the direction of movement. This minimizes the chances of a product being counted as moving in the opposite direction.

2. **Filter by Direction:** Implement a direction filter in the counting logic. Only count objects moving in a specific direction. This can be achieved by analyzing the trajectory of each object and confirming it matches the expected direction before incrementing the count.

3. **Increase Frame Analysis:** Sometimes, increasing the frame rate or the number of frames analyzed can help in more accurately tracking the direction of movement, especially if objects move quickly.

If these adjustments still don't resolve the issue, it might be helpful to look into customizing the tracking algorithm to better suit the specific dynamics of your conveyor setup.

Let us know how it goes or if you need further assistance!
